pandas/tests/frame/methods/test_astype.py

```diff
@@ -726,3 +727,29 @@ def test_astype_categorical_to_string_missing(self):
         cat = df.astype("category")
         result = cat.astype(str)
         tm.assert_frame_equal(result, expected)
+
+
+class IntegerArrayNoCopy(pd.core.arrays.IntegerArray):
+    # GH 42501
+
+    @classmethod
+    def _from_sequence(cls, scalars, *, dtype=None, copy=False):
+        values, mask = coerce_to_array(scalars, dtype=dtype, copy=copy)
+        return IntegerArrayNoCopy(values, mask)
+
+    def copy(self):
+        assert False
+
+
+class Int16DtypeNoCopy(pd.Int16Dtype):
+    # GH 42501
+
+    @classmethod
+    def construct_array_type(cls):
+        return IntegerArrayNoCopy
+
+
+def test_frame_astype_no_copy():
+    # GH 42501
+    df = DataFrame({"col": [1, 4, None, 5]}, dtype=object)
+    df = df.astype({"col": Int16DtypeNoCopy()}, copy=False)
```

Comment:
can you call this result and compare with the original `tm.assert_frame_equal`.

a bit duplicative but readable, can you assert that the arrays inside the result ARE the same as the arrays in the original.

Reply:
I added `assert_frame_equal` and `assert_numpy_array_equal` calls. For the numpy array comparison I used a second column that is not cast. The original column does not (appear to) share memory after the cast -- it's unclear to me what the API guarantees here. Either way, copy as such is no longer called for either column where it was called for both before.

I added an entry under 1.3.2 regressions referencing pd.concat, where the code change is.